src/adc.rs

```diff
@@ -0,0 +1,535 @@
+//! # Work in Progress
+//! API for the ADC (Analog to Digital Converter)
+//!
+//! Note that the more specific your hardware selection is
+//! (e.g. stm32f303**xc** instead of just stm32f303)
+//! the more functionality is accessible
+//! (in this case: ADC3 and ADC4 additionaly to ADC1 and ADC2).
+//!
+//! # Examples
+//! For a simple, working example check `adc.rs` in the examples folder.
+use cortex_m::asm;
+use embedded_hal::adc::{Channel, OneShot};
+
+use crate::rcc::{Clocks, AHB};
+
+#[cfg(feature = "stm32f303")]
+const MAX_ADVREGEN_STARTUP_US: u32 = 10;
+
+#[cfg(feature = "stm32f303")]
+use crate::gpio::{gpioa, gpiob, gpioc, Analog};
+
+#[cfg(any(
+    feature = "stm32f303xb",
+    feature = "stm32f303xc",
+    feature = "stm32f303xd",
+    feature = "stm32f303xe",
+))]
+use crate::gpio::{gpiod, gpioe, gpiof};
+
+#[cfg(feature = "stm32f303")]
+use crate::stm32::{ADC1, ADC1_2, ADC2};
+
+#[cfg(any(
+    feature = "stm32f303xb",
+    feature = "stm32f303xc",
+    feature = "stm32f303xd",
+    feature = "stm32f303xe",
+))]
+use crate::stm32::{ADC3, ADC3_4, ADC4};
+
+/// ADC configuration
+///
+/// TODO: Remove `pub` from the register block once all functionalities are implemented.
+/// Leave it here until then as it allows easy access to the registers.
+pub struct Adc<ADC> {
+    pub rb: ADC,
+    clocks: Clocks,
+    prescale: Prescale,
+    operation_mode: Option<OperationMode>,
+}
+
+/// ADC sampling time
+///
+/// Each channel can be sampled with a different sample time.
+/// For Sampletime T_n the total conversion time (in ADC clock cycles) is
+/// 12.5 + (n + .5)
+///
+/// TODO: there are boundaries on how this can be set depending on the hardware.
+/// Check them and implement a sample time setting mechanism.
+pub enum SampleTime {
+    T_1,
+    T_2,
+    T_4,
+    T_7,
+    T_19,
+    T_61,
+    T_181,
+    T_601,
+}
+
+impl Default for SampleTime {
+    fn default() -> Self {
+        SampleTime::T_19
+    }
+}
+
+impl SampleTime {
+    /// Conversion to bits for SMP
+    fn bitcode(&self) -> u8 {
+        match self {
+            SampleTime::T_1 => 0b000,
+            SampleTime::T_2 => 0b001,
+            SampleTime::T_4 => 0b010,
+            SampleTime::T_7 => 0b011,
+            SampleTime::T_19 => 0b100,
+            SampleTime::T_61 => 0b101,
+            SampleTime::T_181 => 0b110,
+            SampleTime::T_601 => 0b111,
+        }
+    }
+}
+
+#[derive(Clone, Copy, PartialEq)]
+/// ADC operation mode
+///
+/// TODO: Implement other modes (DMA, Differential,…)
+pub enum OperationMode {
+    OneShot,
+}
+
+#[derive(Clone, Copy)]
+/// ADC prescale
+///
+/// TODO: Implement a setter that checks requirements for `HCLK_1` and add it.
+pub enum Prescale {
```

Comment:
As an example for stm32f3 v0.10.0 this enum could be replaced with [CKMODE_W](https://docs.rs/stm32f3/0.10.0/src/stm32f3/stm32f303/adc1_2/ccr.rs.html#328-337).

Reply:
I'd like to postpone this to a later PR when we are resolvint the TODO of testing if rcc_hclk is equal sys_d1cpre_ck.

Reply:
Fair enough

Reply:
Can you explain, what you mean by this, exactly? What is rcc_hclk and what is sy_d1cpre_ck. 

I assume you mean, that rcc_hclk is the clock given by the AHB prescaler and what is sys_d1cpre_ck then? 

Sorry if this question is naiv, but i've difficulties by the naming to find these in the reference manual :)

Reply:
TBH I just copy pasted from the [src that you linked to]:(https://docs.rs/stm32f3/0.10.0/src/stm32f3/stm32f303/adc1_2/ccr.rs.html#331-332)
```
#[doc = "1: Use AHB clock rcc_hclk3. In this case rcc_hclk must equal sys_d1cpre_ck"]
    SYNCDIV1 = 1,
```
As I'm not that good at remembering stuff and wanted to answer quickly I felt like this would be a reliable source for naming conventions. 

Now that I feel guilty for confuing you, lets specify: In [RM0316](https://www.st.com/resource/en/reference_manual/dm00043574.pdf) paragraph 15.3.3. reads:

> Software can use […] CKMODE[1:0]=01 only if the AHB prescaler of the 
RCC is set to 1 (the duty cycle of the AHB clock must be 50% in this configuration).

So that's all I meant by those confusing words above - thanks for tackling me from behind ;)

Reply:
Ah okay. Thank you for the explanation. So this means, that HCLK_1 can only be used, if the AHB prescaler is set to 1. 

So `rcc.cfgr2.read().hpre().is_div1() == true` right?

Reply:
I [implemented the use of `CKMODE`](https://github.com/stm32-rs/stm32f3xx-hal/pull/47/commits/efe8f23ced863e97c18b670a545722d6b82f5557) and I like it more this way. But maybe there is an even more elegant way to go about it?

Using `rcc` is not possible, as it is forzen in `clocks`, so I implemented it differently.
I also added a check to verify the user is not trying to set up ADC1 and ADC2 with different settings (as this is not possible)